File: DTCore5.0-exp/DTMath/DataTools.ExtendedMath/PolarMath.vb
' Polar coordinate translation
' Copyright (C) 2014 Nathaniel Moschkin

' This is the fourth rewriting of this module since 1999.

Option Explicit On
Option Compare Binary
Option Strict On

Imports System.Drawing
Imports DataTools.ExtendedMath.ColorMath

Public Module pMath

    Public Const vbPolarChar = "φ"
    Public Const vbPiChar = "π"
    Public Const vbDegreeChar = "°"
    Public Const vbRadianConst = (180.0# / 3.1415926535897931#)

    Public Enum PolarCoordinateFormatting
        [Default] = &H0

        UseDegreeSymbol = &H1
        UsePolarSymbol = &H2
        UsePiSymbol = &H4
        UseRadianIndicator = &H8

        UseParenthesis = &H10
        UseBrackets = &H20

        DisplayInRadians = &H40

        Standard = &H1
        Radians = &H4 Or &H8 Or &H40

        StandardScientific = &H2
        RadiansScientific = &H2 Or &H4 Or &H8 Or &H40
    End Enum

End Module

Public Class Polar

    Public Enum ColorWheelShapes
        Point = 1
        Hexagon = 2
    End Enum

    Public Structure ColorWheel
        Public Elements() As ColorWheelElement
        Public Bounds As Rectangle
        Public Bitmap() As Byte

        Public Overloads Function HitTest(x As Integer, y As Integer) As System.Drawing.Color
            For Each e As ColorWheelElement In Elements
                For Each f As Point In e.FillPoints
                    If f.X = x And f.Y = y Then Return e.Color
                Next
            Next

            Return Color.Empty
        End Function

        Public Overloads Function HitTest(pt As Point) As System.Drawing.Color
            Return HitTest(pt.X, pt.Y)
        End Function
    End Structure

    Public Structure ColorWheelElement
        Public Color As System.Drawing.Color
        Public Polar As PolarCoordinates
        Public Center As Point
        Public Bounds As Rectangle
        Public FillPoints() As Point
        Public Shape As ColorWheelShapes
    End Structure

    Public Structure PolarCoordinates

        Public Property Formatting As PolarCoordinateFormatting
        Public Property Rounding As Integer
        Public Property Angle As Double
        Public Property Radius As Double

        Public Sub New(radius As Double, angle As Double)
            Formatting = PolarCoordinateFormatting.Standard
            Rounding = 2
            _Radius = radius
            _Angle = angle
        End Sub

        Public Sub New(p As PolarCoordinates)
            Formatting = PolarCoordinateFormatting.Standard
            Rounding = 2
            _Radius = p.Radius
            _Angle = p.Angle
        End Sub

        Public Overrides Function ToString() As String

            Dim s As String = ""
            Dim a As Double = _Angle,
                r As Double = _Radius

            If (Formatting And PolarCoordinateFormatting.UseParenthesis) = PolarCoordinateFormatting.UseParenthesis Then
                s &= "("
            End If

            If (Formatting And PolarCoordinateFormatting.UseBrackets) = PolarCoordinateFormatting.UseBrackets Then
                s &= "{"
            End If

            s &= Format(Math.Round(r, Rounding)) & ", "

            If (Formatting And PolarCoordinateFormatting.DisplayInRadians) = PolarCoordinateFormatting.DisplayInRadians Then
                a *= vbRadianConst
            End If

            s &= Format(Math.Round(a, Rounding))

            If (Formatting And PolarCoordinateFormatting.UsePiSymbol) = PolarCoordinateFormatting.UsePiSymbol Then
                s &= vbPiChar
            End If

            If (Formatting And PolarCoordinateFormatting.UseDegreeSymbol) = PolarCoordinateFormatting.UseDegreeSymbol Then
                s &= vbDegreeChar
            End If

            If (Formatting And PolarCoordinateFormatting.UsePolarSymbol) = PolarCoordinateFormatting.UsePolarSymbol Then
                s &= vbPolarChar
            End If

            If (Formatting And PolarCoordinateFormatting.UseRadianIndicator) = PolarCoordinateFormatting.UseRadianIndicator Then
                s &= " rad"
            End If

            If (Formatting And PolarCoordinateFormatting.UseBrackets) = PolarCoordinateFormatting.UseBrackets Then
                s &= "}"
            End If

            If (Formatting And PolarCoordinateFormatting.UseParenthesis) = PolarCoordinateFormatting.UseParenthesis Then
                s &= ")"
            End If

            Return s

        End Function

        Public Overloads Shared Function rad(p As PolarCoordinates) As UniPoint
            Return rad(p, New Rectangle(0, 0, CInt(p.Radius * 2) + 1, CInt(p.Radius * 2) + 1))
        End Function

        Public Overloads Shared Function rad(p As PolarCoordinates, rect As UniRect) As UniPoint
            If rect.Width < ((p.Radius * 2) + 1) OrElse rect.Height < ((p.Radius * 2) + 1) Then
                '' fit to rectangle
                p = New PolarCoordinates((Math.Min(rect.Width, rect.Height) / 2) - 1, p.Angle)
            End If

            Dim pt As System.Windows.Point = ToScreenCoordinates(p)

            Dim r As Double = p.Radius,
                x As Double,
                y As Double

            Dim c As Point = New Point(CInt(rect.Width / 2), CInt(rect.Height / 2))

            x = c.X + pt.X
            y = c.Y + pt.Y

            pt = New System.Windows.Point(x - 1, y - 1)
            Return New UniPoint(x + rect.Left, y + rect.Top)

        End Function

        Public Shared Widening Operator CType(operand As PolarCoordinates) As System.Windows.Point
            Return ToScreenCoordinates(operand)
        End Operator

        Public Shared Narrowing Operator CType(operand As System.Windows.Point) As PolarCoordinates
            Return ToPolarCoordinates(operand)
        End Operator

    End Structure

    Public Overloads Shared Function ToScreenCoordinates(p As PolarCoordinates) As System.Windows.Point
        Return ToScreenCoordinates(p.Radius, p.Angle)
    End Function

    Public Overloads Shared Function ToScreenCoordinates(r As Double, a As Double) As System.Windows.Point
        Dim x As Double, _
            y As Double

        Dim radConst As Double = (180.0# / 3.1415926535897931#)

        a /= radConst

        y = r * Math.Cos(a)
        x = r * Math.Sin(a)

        Return New System.Windows.Point(x, -y)

    End Function

    Public Overloads Shared Function ToPolarCoordinates(p As System.Windows.Point) As PolarCoordinates
        Return ToPolarCoordinates(p.X, p.Y)
    End Function

    Public Overloads Shared Function ToPolarCoordinates(x As Double, y As Double) As PolarCoordinates

        Dim r As Double, _
            a As Double

        Dim radConst As Double = (180.0# / 3.1415926535897931#)

        r = Math.Sqrt((x * x) + (y * y))

        '' screen coordinates are funny, had to reverse this.
        a = Math.Atan(x / y)

        a *= radConst
        a = (a - 180.0#)
        If a < 0.0# Then a = 360.0# - a
        If a > 360.0# Then a = a - 360.0#

        Return New PolarCoordinates(r, a)

    End Function

    Public Shared Function InWheel(pt As System.Windows.Point, rad As Double) As Boolean
        pt.X -= rad
        pt.Y -= rad
        Dim p As PolarCoordinates = ToPolarCoordinates(pt)

        Return (p.Radius) <= (rad)

    End Function

End Class

File: DTCore5.0-exp/DTMath/DataTools.ExtendedMath/PolarMath.cs
// Polar coordinate translation
// Copyright (C) 2014 Nathaniel Moschkin

// This is the fourth rewriting of this module since 1999.

using System;
using System.Drawing;
using DataTools.ExtendedMath.ColorMath;
using Microsoft.VisualBasic;

namespace DataTools.ExtendedMath
{
    public static class pMath
    {
        public const string vbPolarChar = "φ";
        public const string vbPiChar = "π";
        public const string vbDegreeChar = "°";
        public const double vbRadianConst = 180.0d / 3.1415926535897931d;

        public enum PolarCoordinateFormatting
        {
            Default = 0x0,
            UseDegreeSymbol = 0x1,
            UsePolarSymbol = 0x2,
            UsePiSymbol = 0x4,
            UseRadianIndicator = 0x8,
            UseParenthesis = 0x10,
            UseBrackets = 0x20,
            DisplayInRadians = 0x40,
            Standard = 0x1,
            Radians = 0x4 | 0x8 | 0x40,
            StandardScientific = 0x2,
            RadiansScientific = 0x2 | 0x4 | 0x8 | 0x40
        }
    }

    public class Polar
    {
        public enum ColorWheelShapes
        {
            Point = 1,
            Hexagon = 2
        }

        public struct ColorWheel
        {
            public ColorWheelElement[] Elements;
            public Rectangle Bounds;
            public byte[] Bitmap;

            public Color HitTest(int x, int y)
            {
                foreach (ColorWheelElement e in Elements)
                {
                    foreach (Point f in e.FillPoints)
                    {
                        if (f.X == x & f.Y == y)
                            return e.Color;
                    }
                }

                return Color.Empty;
            }

            public Color HitTest(Point pt)
            {
                return HitTest(pt.X, pt.Y);
            }
        }

        public struct ColorWheelElement
        {
            public Color Color;
            public PolarCoordinates Polar;
            public Point Center;
            public Rectangle Bounds;
            public Point[] FillPoints;
            public ColorWheelShapes Shape;
        }

        public struct PolarCoordinates
        {
            public pMath.PolarCoordinateFormatting Formatting { get; set; }
            public int Rounding { get; set; }
            public double Angle { get; set; }
            public double Radius { get; set; }

            public PolarCoordinates(double radius, double angle)
            {
                Formatting = pMath.PolarCoordinateFormatting.Standard;
                Rounding = 2;
                Radius = radius;
                Angle = angle;
            }

            public PolarCoordinates(PolarCoordinates p)
            {
                Formatting = pMath.PolarCoordinateFormatting.Standard;
                Rounding = 2;
                Radius = p.Radius;
                Angle = p.Angle;
            }

            public override string ToString()
            {
                string s = "";
                double a = Angle;
                double r = Radius;
                if ((Formatting & pMath.PolarCoordinateFormatting.UseParenthesis) == pMath.PolarCoordinateFormatting.UseParenthesis)
                {
                    s += "(";
                }

                if ((Formatting & pMath.PolarCoordinateFormatting.UseBrackets) == pMath.PolarCoordinateFormatting.UseBrackets)
                {
                    s += "{";
                }

                s += Strings.Format(Math.Round(r, Rounding)) + ", ";
                if ((Formatting & pMath.PolarCoordinateFormatting.DisplayInRadians) == pMath.PolarCoordinateFormatting.DisplayInRadians)
                {
                    a *= pMath.vbRadianConst;
                }

                s += Strings.Format(Math.Round(a, Rounding));
                if ((Formatting & pMath.PolarCoordinateFormatting.UsePiSymbol) == pMath.PolarCoordinateFormatting.UsePiSymbol)
                {
                    s += pMath.vbPiChar;
                }

                if ((Formatting & pMath.PolarCoordinateFormatting.UseDegreeSymbol) == pMath.PolarCoordinateFormatting.UseDegreeSymbol)
                {
                    s += pMath.vbDegreeChar;
                }

                if ((Formatting & pMath.PolarCoordinateFormatting.UsePolarSymbol) == pMath.PolarCoordinateFormatting.UsePolarSymbol)
                {
                    s += pMath.vbPolarChar;
                }

                if ((Formatting & pMath.PolarCoordinateFormatting.UseRadianIndicator) == pMath.PolarCoordinateFormatting.UseRadianIndicator)
                {
                    s += " rad";
                }

                if ((Formatting & pMath.PolarCoordinateFormatting.UseBrackets) == pMath.PolarCoordinateFormatting.UseBrackets)
                {
                    s += "}";
                }

                if ((Formatting & pMath.PolarCoordinateFormatting.UseParenthesis) == pMath.PolarCoordinateFormatting.UseParenthesis)
                {
                    s += ")";
                }

                return s;
            }

            public static UniPoint rad(PolarCoordinates p)
            {
                return rad(p, new Rectangle(0, 0, (int)(p.Radius * 2d) + 1, (int)(p.Radius * 2d) + 1));
            }

            public static UniPoint rad(PolarCoordinates p, UniRect rect)
            {
                if (rect.Width < p.Radius * 2d + 1d || rect.Height < p.Radius * 2d + 1d)
                {
                    // ' fit to rectangle
                    p = new PolarCoordinates(Math.Min(rect.Width, rect.Height) / 2d - 1d, p.Angle);
                }

                var pt = ToScreenCoordinates(p);
                double r = p.Radius;
                double x;
                double y;
                var c = new Point((int)(rect.Width / 2d), (int)(rect.Height / 2d));
                x = c.X + pt.X;
                y = c.Y + pt.Y;
                pt = new System.Windows.Point(x - 1d, y - 1d);
                return new UniPoint(x + rect.Left, y + rect.Top);
            }

            public static implicit operator System.Windows.Point(PolarCoordinates operand)
            {
                return ToScreenCoordinates(operand);
            }

            public static explicit operator PolarCoordinates(System.Windows.Point operand)
            {
                return ToPolarCoordinates(operand);
            }
        }

        public static System.Windows.Point ToScreenCoordinates(PolarCoordinates p)
        {
            return ToScreenCoordinates(p.Radius, p.Angle);
        }

        public static System.Windows.Point ToScreenCoordinates(double r, double a)
        {
            double x;
            double y;
            double radConst = 180.0d / 3.1415926535897931d;
            a /= radConst;
            y = r * Math.Cos(a);
            x = r * Math.Sin(a);
            return new System.Windows.Point(x, -y);
        }

        public static PolarCoordinates ToPolarCoordinates(System.Windows.Point p)
        {
            return ToPolarCoordinates(p.X, p.Y);
        }

        public static PolarCoordinates ToPolarCoordinates(double x, double y)
        {
            double r;
            double a;
            double radConst = 180.0d / 3.1415926535897931d;
            r = Math.Sqrt(x * x + y * y);

            // ' screen coordinates are funny, had to reverse this.
            a = Math.Atan(x / y);
            a *= radConst;
            a = a - 180.0d;
            if (a < 0.0d)
                a = 360.0d - a;
            if (a > 360.0d)
                a = a - 360.0d;
            return new PolarCoordinates(r, a);
        }

        public static bool InWheel(System.Windows.Point pt, double rad)
        {
            pt.X -= rad;
            pt.Y -= rad;
            var p = ToPolarCoordinates(pt);
            return p.Radius <= rad;
        }
    }
}